demos/adjoint/adjoint_demo.py

```diff
@@ -0,0 +1,348 @@
+Q = FunctionSpace(mesh, "CG", 2)  # Temperature function space (scalar)
+Z = MixedFunctionSpace([V, W])  # Mixed function space
+
+# Test functions and functions to hold solutions:
+z = Function(Z)  # A field over the mixed function space Z
+u, p = split(z)  # Returns symbolic UFL expression for u and p
+z.subfunctions[0].rename("Velocity")
+z.subfunctions[1].rename("Pressure")
+
+Ra = Constant(1e6)  # Rayleigh number
+approximation = BoussinesqApproximation(Ra)
+
+# Define time-stepping parameters.
+delta_t = Constant(4e-6)  # Constant time step
+timesteps = int(temperature_timestepping_info["index"][-1])  # number of timesteps from forward
+
+T = Function(Q, name="Temperature")
+
+# Nullspaces for the problem (constant pressure).
+Z_nullspace = create_stokes_nullspace(Z, closed=True, rotational=False)
+
+# Stokes and Energy boundary conditions - all free-slip for this problem
+stokes_bcs = {
+    bottom_id: {"uy": 0},
+    top_id: {"uy": 0},
+    left_id: {"ux": 0},
+    right_id: {"ux": 0},
+}
+temp_bcs = {
+    bottom_id: {"T": 1.0},
+    top_id: {"T": 0.0},
+}
+# Energy and Stokes solver
+energy_solver = EnergySolver(T, u, approximation, delta_t, ImplicitMidpoint, bcs=temp_bcs)
+stokes_solver = StokesSolver(z, T, approximation, bcs=stokes_bcs, nullspace=Z_nullspace, transpose_nullspace=Z_nullspace, constant_jacobian=True)
+# -
+
+# Define the Control Space
+# ========================
+#
+# In this section, we define the control space, which can be simplified to reduce the risk of encountering an undetermined problem.
+# Here, we select the Q1 space for the initial condition $T_{ic}$. We also provide an initial guess for the control value, which in this
+# synthetic test is the temperature field of the reference simulation at the final time-step (`timesteps - 1`). In other words, our guess for the initial
+# temperature is the final model state.
+
+# +
+# Define control function space using continuous Galerkin (CG) of degree 1 (Q1)
+Q1 = FunctionSpace(mesh, "CG", 1)
+
+# Create a function for the initial temperature field with a name for identification
+Tic = Function(Q1, name="Initial Temperature")
+
+# Project the temperature field from the reference simulation's final time-step onto the control space as our initial guess
+with CheckpointFile(checkpoint_filename, mode="r") as fi:
+    Tic.project(fi.load_function(mesh, "Temperature", idx=timesteps - 1))
+# -
+
+# We finally introduce our control problem to pyadjoint and continue by integrating the solutions at
+# each time-step. Notice that we cumulatively compute the misfit term with respect to the
+# surface velocity observable.
+
+# +
+control = Control(Tic)
+
+u_misfit = 0.0
+
+# Take our initial guess and project from Q1 to Q2, simultaneously imposing temperature boundary conditions.
+T.project(Tic, bcs=energy_solver.strong_bcs)
+
+# Populate the tape by running the forward simulation.
+# For learning+tweaking purposes, feel free to change e.g., 0 -> timesteps - 3
```

Comment:
I found this hard to underrstand. Maybe provide some more context. For learning and tweaking what?

Reply:
Sorry I should have been more clear here. This is just to say for the purpose of learning, the user can do this exercise for only 5 timesteps. Everything would run fine for that. But in our experiment in Ghelichkhan et al 2024 we had this as 0.